### PR TITLE
ENH add IRLS solver

### PR DESCRIPTION
**Pull request by josephsalmon**

### CODE REVIEW

File: solvers/irls.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    def set_objective(self, X, y, lmbd, fit_intercept):
+        self.X, self.y, self.lmbd = X, y, lmbd
+        self.fit_intercept = fit_intercept
+
+    def run(self, n_iter):
+        epsilon = 1e-25
+
+        def w_opt(eta):
+            T = self.X.T @ self.X + self.lmbd * np.diag(1.0 / eta)
```
Comment (agramfort) [inline solvers/irls.py:35]: You could use the dual formulation when more features than samples. Another option is to scale the X and just use the ridge_solver in sklearn

Comment (josephsalmon) [reply solvers/irls.py]: This now added, thanks for the proposition (v1 was for having the original formula).

File: solvers/irls.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def set_objective(self, X, y, lmbd, fit_intercept):
+        self.X, self.y, self.lmbd = X, y, lmbd
+        self.fit_intercept = fit_intercept
+
+    def run(self, n_iter):
+        epsilon = 1e-25
+
+        def w_opt(eta):
+            ridge = Ridge(alpha=self.lmbd, fit_intercept=False).fit(
+                self.X @ np.diag(eta ** 0.5), self.y
```
Comment (agramfort) [inline solvers/irls.py:41]: Use broadcasting to avoid the allocation of a full diag matrix

**Review by agramfort — COMMENTED** [review]
I would cleanup the remaining commented lines and then LGTM

File: datasets/simulated.py [hunk 3]
```diff
@@ -11,11 +11,8 @@ class Dataset(BaseDataset):
     # List of parameters to generate the datasets. The benchmark will consider
     # the cross product for each key in the dictionary.
     parameters = {
-        'n_samples, n_features': [
-            (100, 5_000),
-            (100, 10_000)
-        ],
-        'rho': [0, 0.6],
+        "n_samples, n_features": [(100, 5_00), (100, 10_0)],
```
Comment (mathurinm) [inline datasets/simulated.py:14]: unintended change

File: datasets/simulated.py [hunk 4]
```diff
@@ -11,11 +11,8 @@ class Dataset(BaseDataset):
...
+        "rho": [0.5],
```
Comment (mathurinm) [inline datasets/simulated.py:15]: same (it will break things in the doc of benchopt, it's a bad design I know.

File: datasets/simulated.py [hunk 5]
```diff
@@ -28,8 +25,9 @@ def __init__(self, n_samples=10, n_features=50, rho=0, random_state=27):
     def get_data(self):
         rng = np.random.RandomState(self.random_state)
 
-        X, y, _ = make_correlated_data(self.n_samples, self.n_features,
-                                       rho=self.rho, random_state=rng)
+        X, y, _ = make_correlated_data(
```
Comment (mathurinm) [inline datasets/simulated.py:28]: we try to avoid formatting change like these which make the diff longer than it could be. some people don't use black :)

File: solvers/irls.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            ridge = Ridge(alpha=self.lmbd, fit_intercept=False).fit(
+                self.X * sqrt_eta, self.y
+            )
+            return ridge.coef_ * sqrt_eta
+
+        def eta_opt(w):
+            return np.abs(w)
+
+        n_features = self.X.shape[1]
+        eta = np.abs(self.X.T @ self.y)  # init needs to be > 0
```
Comment (mathurinm) [inline solvers/irls.py:46]: Can you briefly comment the rationale for this choice (rather than OLS for example ?)

File: solvers/python_pgd.py [hunk 7]
```diff
@@ -6,19 +6,19 @@
 
 
 class Solver(BaseSolver):
-    name = 'Python-PGD'  # proximal gradient, optionally accelerated
+    name = "Python-PGD"  # proximal gradient, optionally accelerated
```
Comment (mathurinm) [inline solvers/python_pgd.py:9]: can you revert these changes ? they are not needed, likely to complicate other PRs, and introduce noise

File: datasets/simulated.py (hunk 3 above)
Comment (josephsalmon) [reply datasets/simulated.py]: sorry, I probably did a commit -am "..." without realising.
Undoing it right now

File: datasets/simulated.py (hunk 4 above)
Comment (josephsalmon) [reply datasets/simulated.py]: idem

File: solvers/irls.py (hunk 6 above)
Comment (josephsalmon) [reply solvers/irls.py:46]: this is a cheap proxy for OLS without doing the computation (OLS solution if X^top X= Id).

File: solvers/python_pgd.py (hunk 7 above)
Comment (josephsalmon) [reply solvers/python_pgd.py]: sure. that was a ctrl + S in vscode that black-ify the file. Sorry about it.

File: solvers/irls.py (hunk 6 above)
Comment (agramfort) [reply solvers/irls.py:46]: Does it bring noticeable improvement on a benchmark?

Comment (josephsalmon) [reply solvers/irls.py]: nope. I simplified, and init is now `np.ones(n_features)` meaning your start in the loop with a vanilla Ridge.

File: solvers/python_pgd.py [hunk 8]
```diff
@@ -10,15 +10,15 @@ class Solver(BaseSolver):
     stopping_strategy = "callback"
 
     # any parameter defined here is accessible as a class attribute
-    parameters = {'use_acceleration': [False, True]}
+    parameters = {"use_acceleration": [False, True]}
```
Comment (agramfort) [inline solvers/python_pgd.py:13]: These changes are also unrelated

Comment (josephsalmon) [reply solvers/python_pgd.py]: I have removed all but the unique file intended here, should solve the problem

**Review by agramfort — APPROVED** [review]

**Comment by mathurinm** [comment]
Thank you @josephsalmon !
